README.md
![image](example/images/reflection_effect2.png)

# Reflection Effect
A Flutter widget that reflects the content of a widget. This widget can be used to create mirror effects or to display information in a more visually appealing way.

## Usage

To use the Reflection widget, you first need to import it into your widget file:

```
import 'package:reflection_effect/reflection_effect.dart';
```

Once you have imported the widget, you can use it as follows:

```
Reflection( 
    child: Text('Hello World',
        style: TextStyle(fontSize: 30, color: Colors.blue)),
)
```

You can also change the reflection setting as follows:
```
    Reflection(
        settingReflection: SettingReflection(
            skewX: 0.2,
            scaleY: 0.5,
            opacity: 0.9,
            reflectionLength: 0.4,
            positionX: 0.2,
            expandRight: 10,
            below: -32,
        ),
        child: const Text('Hello World '),
    ),

```






CHANGELOG.md
## 2.0.0
* settingReflection which allows for better reflection effects.

## 1.0.3
* redame updated and minor issues fixed

## 1.0.2
* fixed problem import in example

## 1.0.1

* add License

## 1.0.0

* initial release.

lib/src/config_reflection.dart
///Create CustomConfigReflection to custom reflection
///The [skewX] is the skew of the reflection 0.0 to 1.0
///The [scaleY] is the size of reflection widget 0.0 to 2.0
///The [opacity] is the opacity of the reflection 0.0 to 1.0
///The [reflectionLength] is the length of the reflection 0.0 to 1.0
///
class SettingReflection {
  double skewX;
  double scaleY;
  double opacity;
  double reflectionLength;
  double positionX;
  double expandRight;
  double below;

  SettingReflection({
    this.skewX = 0.0,
    this.scaleY = 0.90,
    this.opacity = 1.0,
    this.reflectionLength = 8.0,
    this.positionX = 0.0,
    this.expandRight = 0.0,
    this.below = 1.0,
  });
}

enum SettingReflectionEnum {
  text,
  image,
}

extension SettingReflectionEnumExtensionon on SettingReflectionEnum {
  //get
  SettingReflection get setting {
    switch (this) {
      case SettingReflectionEnum.text:
        return SettingReflection(
          skewX: 0.2,
          scaleY: 0.5,
          opacity: 0.9,
          reflectionLength: 0.4,
          positionX: 0.2,
          expandRight: 10,
          below: 10,
        );
      case SettingReflectionEnum.image:
        return SettingReflection(
          skewX: 0.2,
          scaleY: 0.5,
          opacity: 0.5,
          reflectionLength: 0.4,
          positionX: 0.2,
          expandRight: 10,
          below: 1,
        );

      default:
        return SettingReflection();
    }
  }
}


lib/src/reflection_widget.dart
import 'package:flutter/material.dart';
import 'package:reflection_effect/src/config_reflection.dart';

/// Widget that shows a reflection of its child
///
/// The [child] is the widget that will be reflected
/// The [settingReflection] is the setting of the reflection
///
/// Example:
/// ```dart
/// Reflection(
///  settingReflection: SettingReflection(
///     skewX: 0.2,
///     scaleY: 0.5,
///     opacity: 0.9,
///     reflectionLength: 0.4,
///     positionX: 0.2,
///     expandRight: 10,
///     below: -32,
///   child: const Text('Hello World ')
/// ),
///
class Reflection extends StatefulWidget {
  const Reflection({super.key, required this.child, this.settingReflection});

  final Widget child;
  final SettingReflection? settingReflection;

  @override
  State<Reflection> createState() => _ReflectWidgetState();
}

class _ReflectWidgetState extends State<Reflection>
    with WidgetsBindingObserver {
  OverlayPortalController controller = OverlayPortalController();
  Offset childPosition = Offset.zero;
  Size childSize = Size.zero;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
    //Set position and show when widget is ready
    WidgetsBinding.instance.addPostFrameCallback((_) {
      RenderBox childRenderBox = context.findRenderObject() as RenderBox;
      childPosition = childRenderBox.localToGlobal(Offset.zero);
      childSize = childRenderBox.size;
      controller.show();
    });
  }

  //For reload position when size change
  @override
  void didChangeMetrics() async {
    super.didChangeMetrics();
    //Set position went size change
    await Future.delayed(const Duration(milliseconds: 100));
    setState(() {
      RenderBox childRenderBox = context.findRenderObject() as RenderBox;
      childPosition = childRenderBox.localToGlobal(Offset.zero);
      childSize = childRenderBox.size;
    });
  }

  @override
  Widget build(BuildContext context) {
    print(widget.settingReflection);
    SettingReflection sR = widget.settingReflection ?? SettingReflection();

    final double positionX = sR.positionX;
    final double reflectionLength = sR.reflectionLength;
    final double skewX = sR.skewX;
    final double scaleY = sR.scaleY;
    final double expandRight = sR.expandRight;
    final double opacity = sR.opacity;
    final double below = sR.below;

    return OverlayPortal(
      controller: controller,
      overlayChildBuilder: (BuildContext context) {
        return Positioned(
          top: childPosition.dy + childSize.height + below,
          left: childPosition.dx + positionX,
          child: ShaderMask(
            shaderCallback: (Rect bounds) {
              return LinearGradient(
                begin: Alignment.topCenter,
                end: Alignment.bottomCenter,
                colors: const <Color>[
                  Colors.black,
                  Color.fromARGB(0, 0, 0, 0),
                  Color.fromARGB(0, 0, 0, 0)
                ],
                stops: [0.0, reflectionLength, 1.0],
              ).createShader(bounds);
            },
            blendMode: BlendMode.dstIn,
            child: Opacity(
              opacity: opacity,
              child: Transform(
                transform: Matrix4.skewX(skewX),
                child: Transform(
                  transform: Transform.scale(
                    scaleY: scaleY,
                  ).transform,
                  child: Transform(
                    filterQuality: FilterQuality.low,
                    transform: Matrix4.identity()..scale(1.0, -1.0),
                    alignment: Alignment.center,
                    child: Padding(
                      padding: EdgeInsets.only(right: expandRight),
                      child: widget.child,
                    ),
                  ),
                ),
              ),
            ),
          ),
        );
      },
      child: widget.child,
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:reflection_effect/reflection_effect.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Example Reflection Widget',
      home: Scaffold(
        body: Stack(
          children: [
            Center(child: Image.asset('images/backgraund.png')),
            Center(
              child: Row(
                children: [
                  const Spacer(),
                  Reflection(
                    settingReflection: SettingReflection(
                      skewX: 0.2,
                      scaleY: 0.5,
                      opacity: 0.9,
                      reflectionLength: 0.4,
                      positionX: 0.2,
                      expandRight: 10,
                      below: -32,
                    ),
                    child: const Text('Hello World ',
                        style: TextStyle(
                            fontSize: 60,
                            color: Color.fromARGB(255, 15, 35, 90))),
                  ),
                  Reflection(
                    settingReflection: SettingReflectionEnum.image.setting,
                    child: Container(
                      color: Colors.red,
                      height: 50,
                      width: 50,
                    ),
                  ),
                  const Spacer()
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}
